src/Sulu/Bundle/TestBundle/Testing/JsonRequestTrait.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+/*
+ * This file is part of Sulu.
+ *
+ * (c) Sulu GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\TestBundle\Testing;
+
+use Symfony\Bundle\FrameworkBundle\KernelBrowser;
+use Symfony\Component\DomCrawler\Crawler;
+
+trait JsonRequestTrait
+{
+    /**
+     * @param mixed[] $parameters
+     * @param mixed[] $server
+     */
+    protected static function jsonRequest(KernelBrowser $client, string $method, string $uri, array $parameters = [], array $server = [], bool $changeHistory = true): Crawler
```

Comment:
Still not  a huge fan of traits in PHP in general... But apart from that, why is that function `static`? It would be a lot easier to use, because the `$client` would not have to be passed ([other traits also add member variables](https://github.com/sulu/sulu/blob/master/src/Sulu/Bundle/TestBundle/Testing/KernelTrait.php#L22)). And what is the `changeHistory` parameter for?

Reply:
Traits for tests are fine and we use here lot of traits to inject our logic into SuluTestCase but also use them in third party test cases like PantherTestCase.
The method parameters are the same as for the normal request and have the same defaults so there is no bc break and easy to refractor this project by just replace `$this->client->request(` with `static::jsonRequest($this->client, `, think we should make the upgrade for our uses as easy as possible.

Reply:
I know that traits are kind of OK for tests, that's why I am not saying we should remove that trait. But I would still not make the function `static`, because it would be even easier to change. It would basically be replacing `$this->client->request` with `$this->jsonRequest` (or should it be `$this->requestJson`, to have the verb first in the method name?), and it would be easier to call, because there would be a parameter less to pass.

Reply:
I would make it static as we don't know where the method are used sometimes you have things in the setUpBeforeClass and you can only call static function where needed. Please keep in mind this need to be refractored in all our projects this method is not just for us in our tests and I just want to make there the upgrade as easy as possible by doing the replace. As we make it not static this will make it more complicated as it should be and the project query something in static methods can not longer use them.

Reply:
I think there should not be any requests in `setUpBeforeClass`, or are there? Apart from that, this is a newly introduced function, so people have to change their code anyway, it won't work out of the box in any case. Therefore I think it is not worth paying the high price of a more complicated function call.

Actually I think I would not even have introduced the trait at all, but would have fixed the code in places where it is needed (not too many tests were failing). But I can see why you would want to have a trait, but I wouldn't have introduced it, since it just more code we have to maintain. But if we are doing it, I would make it easy to use in the future, and not making it complicated to keep some kind of BC, which is not really there anyway.

Reply:
@danrot

> I think there should not be any requests in setUpBeforeClass, or are there?

Sadly yes there are projects which create fixtures over apis for tests instead of creating them manually over the entity manager.

> not too many tests were failing

But if we look all api tests which sends the json as request parameters instead as content are false test as we don't really test if the Api works really with json content thats why I think it correct to update all tests as I think all the tests are false implemented and not really testing there how the apis are used.

> I would make it easy to use in the future, and not making it complicated to keep some kind of BC

Another method would be is override the `createClient` method to write the `$client` into a private static property and has private `static::getClient` which symfony is using in there BrowserKitAssertionsTrait:

https://github.com/symfony/symfony/blob/ff97b5f17b63d5198cc56584623d1bb09f923b5b/src/Symfony/Bundle/FrameworkBundle/Test/BrowserKitAssertionsTrait.php#L144-L157

sadly we can currently not reuse the getClient of symfony as the method is private so we would need to implement the same behaviour by override the `createClient` method call there a setter or implement something like [symfony does](https://github.com/symfony/symfony/blob/ff97b5f17b63d5198cc56584623d1bb09f923b5b/src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php#L59) which call the [getClient](https://github.com/symfony/symfony/blob/ff97b5f17b63d5198cc56584623d1bb09f923b5b/src/Symfony/Bundle/FrameworkBundle/Test/BrowserKitAssertionsTrait.php#L144-L157) but this did look for me hackier then just give the client to the method.

Reply:
Sounds more and more to me like we should fix that properly and use the client as Symfony intends it instead of creating a new abstraction that we have to maintain, is harder to understand and debug, and might lead to some problems in the future, especially if we also want to keep BC for that.

But if I am the only person thinking like that I am happy to be outvoted by @sulu/core-team.

Reply:
I have just seen that you have [contributed a similar function in Symfony](https://github.com/symfony/symfony/pull/38596/files). Why not implementing it now in a way that it is compatible with that one, so that we can delete our code once we don't need it anymore? Even if it is more hacky (not sure if that is the case, since I don't know the implementation details) I would prefer that, because it is less likely to stay in our codebase than the current solution.

Reply:
I've just had a deeper look, and I think all we have to do in order to make the new `jsonRequest` function work you've introduced in your Symfony PR, is to decorate the `test.client` service from symfony, because that seems to be what the `createClient` method uses: https://github.com/symfony/symfony/blob/4.3/src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php#L46

Then we can make use of that new function everywhere, and delete the code we copied from your Symfony PR (or maybe even the entire decorator) as soon as we only support Symfony versions containing this function. Or am I missing something?